In [1]:
# Importing the dataset

import pandas as pd

path = "/Users/rohityadav/Desktop/Git Projects/ml-credit-risk-modeling/data/train.csv"
df = pd.read_csv(path)
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [2]:
# Exploring the dataset

df.info()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             601 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [3]:
df["Loan_Status"].value_counts(normalize = True)

Loan_Status
Y    0.687296
N    0.312704
Name: proportion, dtype: float64

---------------------------------------------------------------------------------------------------------------------------------------------

**Credit history vs Loan Status**

If someone has paid loans properly before, they’re trusted more.

So we expect:

    Good credit history → more approvals

    Bad or missing credit history → more rejections

If this isn’t true, that’s shocking.

Credit_History is a proxy for prior repayment behavior and is typically the strongest univariate predictor in credit risk datasets.

We examine conditional approval/default rates.

---------------------------------------------------------------------------------------------------------------------------------------------

In [4]:
df.groupby('Credit_History')['Loan_Status'].value_counts(normalize = True)

Credit_History  Loan_Status
0.0             N              0.921348
                Y              0.078652
1.0             Y              0.795789
                N              0.204211
Name: proportion, dtype: float64

---------------------------------------------------------------------------------------------------------------------------------------------

**What does this tell us?**

People with a missing credit history (0.0) have a very high rejection rate (0.921 ~ 92%) and a very low success rate (0.0786 ~ 7.8%).
People with a credit history have a mixed proportion with a moderate high success rate (0.795 ~ 79.5%) and a moderate rejection rate (0.204 ~ 20.4%).

---------------------------------------------------------------------------------------------------------------------------------------------

---------------------------------------------------------------------------------------------------------------------------------------------

**Loan Amount vs Loan Status**

Loan amount represents exposure:

    Bigger loans → higher potential loss

Banks are more cautious

---------------------------------------------------------------------------------------------------------------------------------------------

In [10]:
df.groupby('LoanAmount')['Loan_Status'].value_counts(normalize = True)

LoanAmount  Loan_Status
9.0         N              1.0
17.0        Y              1.0
25.0        Y              1.0
26.0        Y              1.0
30.0        Y              1.0
                          ... 
500.0       N              1.0
570.0       N              1.0
600.0       Y              1.0
650.0       Y              1.0
700.0       Y              1.0
Name: proportion, Length: 280, dtype: float64

In [11]:
df.groupby('Loan_Status')['LoanAmount'].median()

Loan_Status
N    129.0
Y    126.0
Name: LoanAmount, dtype: float64

---------------------------------------------------------------------------------------------------------------------------------------------

**What does this tell us?**

Rejected loans typically have higher loan amounts.

---------------------------------------------------------------------------------------------------------------------------------------------

---------------------------------------------------------------------------------------------------------------------------------------------

**Income vs Loan Status**

Income seems important intuitively, but raw income is often noisy.

What we will be checking for?

Median applicant and coapplicant income by loan status

---------------------------------------------------------------------------------------------------------------------------------------------

In [15]:
df.groupby('Loan_Status')[['ApplicantIncome', 'CoapplicantIncome']].median()

ApplicantIncome  CoapplicantIncome
Loan_Status                                    
N                     3833.5              268.0
Y                     3812.5             1239.5

---------------------------------------------------------------------------------------------------------------------------------------------

**What does this tell us?**

Income differences are often small or inconsistent

Confirms income alone is not enough to judge risk

---------------------------------------------------------------------------------------------------------------------------------------------

---------------------------------------------------------------------------------------------------------------------------------------------

**Categorical Feature: Property Area vs Loan Status**

Categorical features often act as socioeconomic proxies.

What we check

Approval/rejection rates by property area

---------------------------------------------------------------------------------------------------------------------------------------------

In [18]:
df.groupby('Property_Area')['Loan_Status'].value_counts(normalize = True)

Property_Area  Loan_Status
Rural          Y              0.614525
               N              0.385475
Semiurban      Y              0.768240
               N              0.231760
Urban          Y              0.658416
               N              0.341584
Name: proportion, dtype: float64

---------------------------------------------------------------------------------------------------------------------------------------------

**What does this tell us?**

There is very slight difference across different categories - Rural, Semiurban, Urban.

This criteria is useful but not dominant.

---------------------------------------------------------------------------------------------------------------------------------------------